### PR TITLE
Add `Message`, setup `Event` being a kind of `Message`

### PR DESCRIPTION
**Pull request by jcredding**
This adds a base `Message` class, commonizes logic from the `Job`
class and updates `Job` to inherit from it. This is setup for
`Event` no longer composing a job and instead inheriting from
message.

This is part of a larger effort to remove some of the strange
behavior with events composing jobs. The goal is that both jobs
and events are handled the same when routing them to a handler
but they can be distinct in how they handle their name, params and
other data. We currently have consistent handling with routing but
have awkward logic related to making events and jobs distinct. By
making event a message it can be a distinct class but be routed
consistently as a generic message.

Every kind of message has a payload type, route id and params.
These are the common pieces of data that all messages need to
provide so they can added to queues and routed to handlers. Each
message must provide a route name for its route id.

This also renames many uses of `job` to `message` where the object
may be either an event or job and route name to route id. This is
to be explicit about what the object represents and how it can be
used.

@kellyredding - Ready for review.

### CODE REVIEW

**Comment by kellyredding** [comment]
@jcredding we discussed:

* just passing type and params to message constructor
* no implementing a route name method
* lazy-using the route name method to build a route id (renaming to route id throughout the code)

I'll look again after that is addressed.

**Comment by jcredding** [comment]
@kellyredding - I updated this with the changes we discussed. I also renamed usage of route name to route id where applicable. This includes `Route` now having an `id` instead of a `name`.

File: lib/qs/job.rb [hunk 1]
```diff
@@ -32,10 +33,9 @@ def inspect
 
     def ==(other)
       if other.kind_of?(self.class)
-        self.payload_type == other.payload_type &&
```
Comment (kellyredding) [inline lib/qs/job.rb]: @jcredding why did you remove payload type?

File: lib/qs/payload_handler.rb [hunk 2]
```diff
@@ -70,12 +70,12 @@ def raise_if_debugging!(exception)
...
-      log_verbose "===== Running job ====="
+      log_verbose "===== Received message ====="
     end
 
-    def log_job(job)
-      log_verbose "  Job:     #{job.name.inspect}"
-      log_verbose "  Params:  #{job.params.inspect}"
+    def log_message(message)
+      log_verbose "  Name:    #{message.route_id.inspect}"
+      log_verbose "  Params:  #{message.params.inspect}"
```
Comment (kellyredding) [inline lib/qs/payload_handler.rb]: @jcredding didn't you want to do some specific logging depending on whether it was a job or event message?

File: lib/qs/payload_handler.rb [hunk 3]
```diff
@@ -88,9 +88,9 @@ def log_complete(redis_item)
         'time'    => redis_item.time_taken,
         'handler' => redis_item.handler_class
       }
-      if (job = redis_item.job)
-        summary_line_args['job']    = job.name
-        summary_line_args['params'] = job.params
+      if (message = redis_item.message)
+        summary_line_args['name']   = message.route_id
```
Comment (kellyredding) [inline lib/qs/payload_handler.rb]: @jcredding why are you calling the arg `name

File: lib/qs/payload_handler.rb (hunk 2 above)
Comment (kellyredding) [inline lib/qs/payload_handler.rb]: @jcredding by this I mean I thought you wanted to log specific things on events.

File: lib/qs/payload_handler.rb [hunk 4]
```diff
@@ -88,9 +88,9 @@ def log_complete(redis_item)
...
+        summary_line_args['params'] = message.params
```
Comment (kellyredding) [inline lib/qs/payload_handler.rb]: @jcredding it is cool that we can just log params straight up and it works whether the message is a job or event.

File: test/support/factory.rb [hunk 5]
```diff
@@ -13,6 +13,10 @@ def self.exception(klass = nil, message = nil)
     exception
   end
 
+  def self.message(params = nil)
+    self.send([:job, :event_job].choice, params)
```
Comment (kellyredding) [inline test/support/factory.rb]: @jcredding this will change to just `:event` in a future PR, right?

**Comment by kellyredding** [comment]
@jcredding a few more questions - I'll look again once you answer them - thanks!

File: lib/qs/job.rb (hunk 1 above)
Comment (jcredding) [inline lib/qs/job.rb]: @kellyredding - This is jumping the gun a little bit but once event no longer composes job this isn't needed. The `kind_of` check will cover checking for the payload type because all `Job` payload type will be the same.

I can restore it for now.

File: lib/qs/payload_handler.rb (hunk 2 above)
Comment (jcredding) [inline lib/qs/payload_handler.rb]: @kellyredding - I wouldn't mind doing specific logging but I was thinking it would require a conditional and didn't want to do that.

I was just thinking though that I could do like the payload module and use `send` here. Then I could have `log_job` and `log_event`. This would only work for the verbose logging and not the summary logging though.

This would allow the following for events:

```
Channel: a-channel
Name:    an-event
Params:  ...
```

I can make part of this change now adding the `send`, `log_job` and `log_event` methods. The event method will temporarily be logging an event job not an event which will have to be changed once its an actual event. Or I can make the change later once event is a kind of message.

File: lib/qs/payload_handler.rb (hunk 3 above)
Comment (jcredding) [inline lib/qs/payload_handler.rb]: @kellyredding - The key is what is displayed in the summary line. So it will show `name="job|build_pdf"`. It can be `route-id` but I wasn't sure the basic user should be aware of route ids and see those in their logs.

File: test/support/factory.rb (hunk 5 above)
Comment (jcredding) [inline test/support/factory.rb]: @kellyredding - Yes

File: lib/qs/job.rb (hunk 1 above)
Comment (kellyredding) [inline lib/qs/job.rb]: @jcredding no, that's OK - I'm good with this.  Thanks for discussing.

File: lib/qs/payload_handler.rb (hunk 2 above)
Comment (kellyredding) [inline lib/qs/payload_handler.rb]: @jcredding like we discussed, after an event subclasses `Message` formally, I'd like to come in here and do event/job specific logging (both verbose and summary).

File: lib/qs/payload_handler.rb (hunk 3 above)
Comment (kellyredding) [inline lib/qs/payload_handler.rb]: @jcredding yeah, I agree on exposing route id in the logs.  We discussed handling this by doing event/job specific logging in a later effort.

**Comment by kellyredding** [comment]
@jcredding I'm good with this - thanks for the discussion :-1:
